### PR TITLE
TYP: core/dtypes/cast.py

### PR DESCRIPTION
**Pull request by arw2019**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -181,7 +181,7 @@ def maybe_downcast_to_dtype(result, dtype):
     return result
 
 
-def maybe_downcast_numeric(result, dtype, do_round: bool = False):
+def maybe_downcast_numeric(result, dtype: Dtype, do_round: bool = False):
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:184]: DtypeObj?

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -324,7 +324,9 @@ def maybe_cast_result_dtype(dtype: DtypeObj, how: str) -> DtypeObj:
     return dtype
 
 
-def maybe_cast_to_extension_array(cls: Type["ExtensionArray"], obj, dtype=None):
+def maybe_cast_to_extension_array(
+    cls: Type["ExtensionArray"], obj, dtype: Dtype = None
+):
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:329]: docstring says ExtensionDtype, optional, so `Optional[ExtensionDtype]`?

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -722,7 +726,9 @@ def infer_dtype_from_scalar(val, pandas_dtype: bool = False) -> Tuple[DtypeObj,
 
 
 # TODO: try to make the Any in the return annotation more specific
-def infer_dtype_from_array(arr, pandas_dtype: bool = False) -> Tuple[DtypeObj, Any]:
+def infer_dtype_from_array(
+    arr, pandas_dtype: bool = False
+) -> Tuple[DtypeObj, AnyArrayLike]:
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:731]: i think ArrayLike, not AnyArrayLike

File: pandas/core/dtypes/cast.py [hunk 4]
```diff
@@ -996,7 +1016,9 @@ def astype_nansafe(arr, dtype, copy: bool = True, skipna: bool = False):
     return arr.view(dtype)
 
 
-def maybe_convert_objects(values: np.ndarray, convert_numeric: bool = True):
+def maybe_convert_objects(
+    values: np.ndarray, convert_numeric: bool = True
+) -> Union[np.ndarray, ABCDatetimeIndex]:
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:1021]: use "DatetimeIndex" instead of ABCDatetimeIndex

File: pandas/core/dtypes/cast.py [hunk 5]
```diff
@@ -1183,7 +1205,7 @@ def convert_dtypes(
     return inferred_dtype
 
 
-def maybe_castable(arr) -> bool:
+def maybe_castable(arr: np.ndarray) -> bool:
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:1208]: we never get EAs here?

Comment (arw2019) [reply pandas/core/dtypes/cast.py:1210]: Atm the only use case is here:
https://github.com/pandas-dev/pandas/blob/126f3cc7e92db0dd92356269d4fb6ea00db30278/pandas/core/construction.py#L538-L557
which according to docstring excludes `ExtensionArray`

File: pandas/core/dtypes/cast.py (hunk 4 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py]: Done

File: pandas/core/dtypes/cast.py (hunk 3 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py]: Done

File: pandas/core/dtypes/cast.py (hunk 2 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py]: Done

File: pandas/core/dtypes/cast.py (hunk 1 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py]: Switched to that

File: pandas/core/dtypes/cast.py [hunk 6]
```diff
@@ -996,7 +1018,9 @@ def astype_nansafe(arr, dtype, copy: bool = True, skipna: bool = False):
     return arr.view(dtype)
 
 
-def maybe_convert_objects(values: np.ndarray, convert_numeric: bool = True):
+def maybe_convert_objects(
+    values: np.ndarray, convert_numeric: bool = True
+) -> Union[np.ndarray, DatetimeIndex]:
```
Comment (arw2019) [inline pandas/core/dtypes/cast.py:1023]: Getting a circular import here with `DatetimeIndex`. 

Will revert to `ABCDatetimeIndex` for now and circle back at the end

Comment (arw2019) [reply pandas/core/dtypes/cast.py:1023]: actually nvm https://github.com/pandas-dev/pandas/pull/36842#pullrequestreview-506058423 solves this

File: pandas/core/dtypes/cast.py [hunk 7]
```diff
@@ -118,7 +131,7 @@ def is_nested_object(obj) -> bool:
     return False
 
 
-def maybe_downcast_to_dtype(result, dtype):
+def maybe_downcast_to_dtype(result, dtype: Dtype):
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:134]: should this be DtypeObj?

File: pandas/core/dtypes/cast.py [hunk 8]
```diff
@@ -904,7 +936,7 @@ def maybe_upcast(values, fill_value=np.nan, dtype=None, copy: bool = False):
     return values, fill_value
 
 
-def invalidate_string_dtypes(dtype_set):
+def invalidate_string_dtypes(dtype_set: Set):
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:939]: Set[DtypeObj] I think?

File: pandas/core/dtypes/cast.py (hunk 5 above)
Comment (jreback) [reply pandas/core/dtypes/cast.py:1286]: can you add an assert in the body itself to lock this down. (whether we want it or not is different, but guarantee is good for now)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
cc @simonjayhawkins

File: pandas/core/dtypes/cast.py (hunk 7 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py:134]: has to be `Dtype` I think because could be a `str`

File: pandas/core/dtypes/cast.py (hunk 8 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py]: Done

File: pandas/core/dtypes/cast.py (hunk 5 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py:1286]: Added

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@simonjayhawkins i fyou can have a quick look and merge if good

File: pandas/core/dtypes/cast.py [hunk 9]
```diff
@@ -329,7 +342,9 @@ def maybe_cast_result_dtype(dtype: DtypeObj, how: str) -> DtypeObj:
     return dtype
 
 
-def maybe_cast_to_extension_array(cls: Type["ExtensionArray"], obj, dtype=None):
+def maybe_cast_to_extension_array(
+    cls: Type["ExtensionArray"], obj: ArrayLike, dtype: Optional[ExtensionDtype] = None
+) -> ArrayLike:
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/cast.py:347]: `ArrayLike` is a TypeVar. so the return type is typed to be the same as obj.

I _think_ mypy is accepting this since np.ndarray resolves to `Any`

We probably need an ArrayLikeUnion, but this is perhaps out of scope here as the discussion has been started elsewhere. xref  #36100

File: pandas/core/dtypes/cast.py [hunk 10]
```diff
@@ -468,7 +487,7 @@ def maybe_casted_values(index, codes=None):
 
     # if we have the codes, extract the values with a mask
     if codes is not None:
-        mask = codes == -1
+        mask: np.ndarray = codes == -1
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/cast.py:490]: why was this added. mypy is not reporting any errors for me with this removed.

generally, try to avoid adding variable type annotations. should be inferred from expression.

File: pandas/core/dtypes/cast.py (hunk 9 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py:347]: Shall I remove the annotation and add a TODO? 

As an aside FWIW I like the `ArrayLike`/`ArrayLikeT` idea in https://github.com/pandas-dev/pandas/pull/36100/files#r483753222

File: pandas/core/dtypes/cast.py [hunk 11]
```diff
@@ -1184,7 +1220,7 @@ def soft_convert_objects(
 
 
 def convert_dtypes(
-    input_array,
+    input_array: AnyArrayLike,
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/cast.py:1223]: Can you update the docstring to match?

File: pandas/core/dtypes/cast.py (hunk 10 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py:490]: mistake, will revert

I had tried typing `index` and `codes` but there were a lot of errors - this was part of an attempted fix.

File: pandas/core/dtypes/cast.py [hunk 12]
```diff
@@ -1566,7 +1606,9 @@ def find_common_type(types: List[DtypeObj]) -> DtypeObj:
     return np.find_common_type(types, [])
 
 
-def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.ndarray:
+def cast_scalar_to_array(
+    shape: Tuple, value: Scalar, dtype: Optional[DtypeObj] = None
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/cast.py:1610]: we just pass shape onto np.empty, https://numpy.org/doc/1.19/reference/generated/numpy.empty.html

so could be int or tuple of int

I recall a discussion about adding an alias to pandas._typing or

```suggestion
    shape: Union[int, Tuple[int, ...]], value: Scalar, dtype: Optional[DtypeObj] = None
```

and the docstring would need to be updated too.

File: pandas/core/dtypes/cast.py (hunk 11 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py:1225]: Done

File: pandas/core/dtypes/cast.py (hunk 12 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py]: Pushed the suggestion.

Adding an alias to _typing sounds like a good idea. Do we already have a PR on that or shall I submit one?

File: pandas/core/dtypes/cast.py (hunk 9 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/cast.py:349]: you can leave as is for now. we have #36092 to track the issues once np.ndarray no longer resolves to Any.

File: pandas/core/dtypes/cast.py (hunk 12 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/cast.py]: > Do we already have a PR on that or shall I submit one?

I don't think so, go ahead.

File: pandas/core/dtypes/cast.py [hunk 13]
```diff
@@ -1698,7 +1742,7 @@ def construct_1d_ndarray_preserving_na(
     return subarr
 
 
-def maybe_cast_to_integer_array(arr, dtype, copy: bool = False):
+def maybe_cast_to_integer_array(arr, dtype: Union[str, np.dtype], copy: bool = False):
```
Comment (simonjayhawkins) [inline pandas/core/dtypes/cast.py:1745]: is this limited to numpy dtypes? if not should use Dtype from typing.

**Comment by simonjayhawkins** [comment]
Thanks @arw2019 lgtm one more comment https://github.com/pandas-dev/pandas/pull/37024#discussion_r503495620 ping when green

File: pandas/core/dtypes/cast.py (hunk 13 above)
Comment (arw2019) [reply pandas/core/dtypes/cast.py]: it's not. Switched to Dtype here

**Comment by jreback** [comment]
thanks @arw2019
